### ISSUE TITLE
typeahead buffer consumption blocked in "gui.c" versions of vim when there are no GUI events in message queue

### ISSUE THREAD

**Issue by raymond-w-ko** [vim/vim#1746]
Hello all,

I am coding my own async auto-completion plugin. It uses timers in `TextChangedI` autocmd to trigger a computation function after some time. When another key is typed before the timer is complete, it cancels that timer. When you rest, then the computation function triggers. If there are completions, then `feedkeys("<Plug>CompletionTrigger")` to trigger a `completefunc`. The `<Plug>` is defined as `inoremap <silent> <Plug>CompletorTrigger <c-x><c-u><c-p>`.

The main triggering code is almost a straight copy from  https://github.com/maralla/completor.vim/blob/master/autoload/completor.vim

This works fine in terminal version of vim (tested on Linux, cygwin64, Bash on Ubuntu on Windows) but does not work well in GUI versions of vim which uses the `gui.c` backend. I was doing some debugging and I think I have narrowed down the cause. Note that this is hard to track as it is timing related, and other things like having the Task Manager open, or moving the mouse, or pressing the "Control" key causes it to work correctly by injecting a spurious GUI event and the update loop to tick. 

First some assumptions:
* Turn blinking off. Prevents any `WM_TIMER` from causing wakeup.
* Set `updatetime` to be very high like `99999999`. This prevents `CursorHoldI` events.
* Ensure Task Manager is not running. It actually injects `WM_MOUSEMOVE` to every process every tick (https://stackoverflow.com/questions/17818493/mousemove-event-repeating-every-second). This is not a joke.

Here is how I think the code is flowing:
* The timer fires in Insert mode.
* Inside the timer callback, `feedkeys()` is used to inject `<Plug>CompletionTrigger`.
* GUI vim yields.
* Now assuming all the assumptions above listed is true, then main loop is blocked waiting for the next event.
  * Moving the mouse, or tapping "Ctrl" wakes up the loop and causes the typeahead buffer to be processed, and only then does the popup menu appear.

How I think the C code is working:
In `gui.c` at `gui_wait_for_chars(long wtime)`  it reaches this stage
```
    /*
     * We may want to trigger the CursorHold event.  First wait for
     * 'updatetime' and if nothing is typed within that time put the
     * K_CURSORHOLD key in the input buffer.
     */
    if (gui_wait_for_chars_or_timer(p_ut) == OK)
	retval = OK;
#ifdef FEAT_AUTOCMD
```

Then in `gui.c` at `gui_wait_for_chars_or_timer(long wtime)`, it detects that the `typebuf` has changed and returns `FALSE`.
```
    while (wtime < 0 || remaining > 0)
    {
	/* Trigger timers and then get the time in wtime until the next one is
	 * due.  Wait up to that time. */
	due_time = check_due_timer();
	if (typebuf.tb_change_cnt != tb_change_cnt)
	{
	    /* timer may have used feedkeys() */
	    return FALSE;
	}
```

At this point, everything seems correct, my function is timer, and it did use `feedkeys()`

However, once it returns `FALSE`, it then enters into blocking wait.
```
#endif

    if (retval == FAIL)
    {
	/* Blocking wait. */
	before_blocking();
	retval = gui_wait_for_chars_or_timer(-1L);
    }

    gui_mch_stop_blink();
    return retval;
}
```

When the `gui_wait_for_chars_or_timer()` is called again, there is no timers, since it has already fired, and `typebuf.tb_change_cnt == tb_change_cnt` so it does not return `FALSE`.

I feel this is incorrect. I think there should be a check for the `typebuf` before entering into blocking wait. I'm not sure how best to accomplish this. Should be there be:
* another return value to say there is something immediately pending? I noticed that `gui_wait_for_chars_or_timer()` is returning `TRUE` and `FALSE`, yet `gui_wait_for_chars()` is checking against `OK` and `FAIL`. Is this okay?
* Move `tb_change_cnt` to be in the outside function so that `gui_wait_for_chars_or_timer()` uses the same one via a pointer and always trigger the `return FALSE;` statement?
* An explicit `typebuf` before blocking and returns if there is something in the typeahead buffer?

**Comment by brammool** [comment]
Raymond Ko wrote:

> Hello all,
> 
> I am coding my own async auto-completion plugin. It uses timers in `TextChangedI` autocmd to trigger a computation function after some time. When another key is typed before the timer is complete, it cancels that timer. When you rest, then the computation function triggers. If there are completions, then `feedkeys("<Plug>CompletionTrigger")` to trigger a `completefunc`. The `<Plug>` is defined as `inoremap <silent> <Plug>CompletorTrigger <c-x><c-u><c-p>`.
> 
> The main triggering code is almost a straight copy from  https://github.com/maralla/completor.vim/blob/master/autoload/completor.vim
> 
> This works fine in terminal version of vim (tested on Linux, cygwin64, Bash on Ubuntu on Windows) but does not work well in GUI versions of vim which uses the `gui.c` backend. I was doing some debugging and I think I have narrowed down the cause. Note that this is hard to track as it is timing related, and other things like having the Task Manager open, or moving the mouse, or pressing the "Control" key causes it to work correctly by injecting a spurious GUI event and the update loop to tick. 
> 
> First some assumptions:
> * Turn blinking off. Prevents any `WM_TIMER` from causing wakeup.
> * Set `updatetime` to be very high like `99999999`. This prevents `CursorHoldI` events.
> * Ensure Task Manager is not running. It actually injects `WM_MOUSEMOVE` to every process every tick (https://stackoverflow.com/questions/17818493/mousemove-event-repeating-every-second). This is not a joke.
> 
> Here is how I think the code is flowing:
> * The timer fires in Insert mode.
> * Inside the timer callback, `feedkeys()` is used to inject `<Plug>CompletionTrigger`.
> * GUI vim yields.
> * Now assuming all the assumptions above listed is true, then main loop is blocked waiting for the next event.
>   * Moving the mouse, or tapping "Ctrl" wakes up the loop and causes the typeahead buffer to be processed, and only then does the popup menu appear.
> 
> How I think the C code is working:
> In `gui.c` at `gui_wait_for_chars(long wtime)`  it reaches this stage
> ```
>     /*
>      * We may want to trigger the CursorHold event.  First wait for
>      * 'updatetime' and if nothing is typed within that time put the
>      * K_CURSORHOLD key in the input buffer.
>      */
>     if (gui_wait_for_chars_or_timer(p_ut) == OK)
> 	retval = OK;
> #ifdef FEAT_AUTOCMD
> ```
> 
> Then in `gui.c` at `gui_wait_for_chars_or_timer(long wtime)`, it detects that the `typebuf` has changed and returns `FALSE`.
> ```
>     while (wtime < 0 || remaining > 0)
>     {
> 	/* Trigger timers and then get the time in wtime until the next one is
> 	 * due.  Wait up to that time. */
> 	due_time = check_due_timer();
> 	if (typebuf.tb_change_cnt != tb_change_cnt)
> 	{
> 	    /* timer may have used feedkeys() */
> 	    return FALSE;
> 	}
> ```
> 
> At this point, everything seems correct, my function is timer, and it did use `feedkeys()`
> 
> However, once it returns `FALSE`, it then enters into blocking wait.
> ```
> #endif
> 
>     if (retval == FAIL)
>     {
> 	/* Blocking wait. */
> 	before_blocking();
> 	retval = gui_wait_for_chars_or_timer(-1L);
>     }
> 
>     gui_mch_stop_blink();
>     return retval;
> }
> ```
> 
> When the `gui_wait_for_chars_or_timer()` is called again, there is no timers, since it has already fired, and `typebuf.tb_change_cnt == tb_change_cnt` so it does not return `FALSE`.
> 
> I feel this is incorrect. I think there should be a check for the `typebuf` before entering into blocking wait. I'm not sure how best to accomplish this. Should be there be:
> * another return value to say there is something immediately pending? I noticed that `gui_wait_for_chars_or_timer()` is returning `TRUE` and `FALSE`, yet `gui_wait_for_chars()` is checking against `OK` and `FAIL`. Is this okay?

That's a mistake, but the values should be OK.

> * Move `tb_change_cnt` to be in the outside function so that `gui_wait_for_chars_or_timer()` uses the same one via a pointer and always trigger the `return FALSE;` statement?
> * An explicit `typebuf` before blocking and returns if there is something in the typeahead buffer?

It should be similar to what mch_inchar() does in os_unix.c.
Does this solve your problem:

*** ../vim-8.0.0617/src/gui.c	2017-03-16 19:58:19.420253384 +0100
--- src/gui.c	2017-06-04 22:44:00.609571377 +0200
***************
*** 2849,2854 ****
--- 2849,2858 ----
      }
  }
  
+ /*
+  * Returns OK if a character was found to be available within the given time,
+  * or FAIL otherwise.
+  */
      static int
  gui_wait_for_chars_or_timer(long wtime)
  {
***************
*** 2869,2884 ****
  	if (typebuf.tb_change_cnt != tb_change_cnt)
  	{
  	    /* timer may have used feedkeys() */
! 	    return FALSE;
  	}
  	if (due_time <= 0 || (wtime > 0 && due_time > remaining))
  	    due_time = remaining;
  	if (gui_mch_wait_for_chars(due_time))
! 	    return TRUE;
  	if (wtime > 0)
  	    remaining -= due_time;
      }
!     return FALSE;
  #else
      return gui_mch_wait_for_chars(wtime);
  #endif
--- 2873,2888 ----
  	if (typebuf.tb_change_cnt != tb_change_cnt)
  	{
  	    /* timer may have used feedkeys() */
! 	    return FAIL;
  	}
  	if (due_time <= 0 || (wtime > 0 && due_time > remaining))
  	    due_time = remaining;
  	if (gui_mch_wait_for_chars(due_time))
! 	    return OK;
  	if (wtime > 0)
  	    remaining -= due_time;
      }
!     return FAIL;
  #else
      return gui_mch_wait_for_chars(wtime);
  #endif
***************
*** 2896,2901 ****
--- 2900,2906 ----
  gui_wait_for_chars(long wtime)
  {
      int	    retval;
+     int	    tb_change_cnt = typebuf.tb_change_cnt;
  
  #ifdef FEAT_MENU
      /*
***************
*** 2953,2959 ****
      }
  #endif
  
!     if (retval == FAIL)
      {
  	/* Blocking wait. */
  	before_blocking();
--- 2958,2964 ----
      }
  #endif
  
!     if (retval == FAIL && typebuf.tb_change_cnt == tb_change_cnt)
      {
  	/* Blocking wait. */
  	before_blocking();

-- 
A)bort, R)etry, B)ang it with a large hammer

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by raymond-w-ko** [comment]
Yes, this works! Thank you! 😄 

Here is the patch to those curious:
https://github.com/raymond-w-ko/vim/commit/3a0fc3e11b16da362426cf1ec644c92140406add

**Comment by raymond-w-ko** [comment]
Closed via commit 9472eec83c3f9c191814dc81dd82498c10b1fc9c
